src/main/java/li/cil/oc2/common/inet/DefaultSessionLayer.java

```diff
@@ -0,0 +1,178 @@
+
+    ///////////////////////////////////////////////////////////////////
+
+    private static final Executor executor = Executors
+            .newSingleThreadExecutor(runnable -> new Thread(runnable, "internet/blocking-session"));
+
+    private static final Selector selector;
+
+    ///////////////////////////////////////////////////////////////////
+
+    static {
+        Selector newSelector;
+        try {
+            newSelector = Selector.open();
+        } catch (IOException e) {
+            LOGGER.error("Failed to open selector", e);
+            newSelector = null;
+        }
+        selector = newSelector;
+        InternetManager.registerNetworkThreadAction(DefaultSessionLayer::selectAction);
+    }
+
+    private final AtomicReference<EchoResponse> echoResponse = new AtomicReference<>(null);
+
+    ///////////////////////////////////////////////////////////////////
+    private final Set<SelectionKey> datagramKeys = new HashSet<>();
+
+    private static void selectAction() {
+        try {
+            selector.selectNow();
+        } catch (IOException e) {
+            LOGGER.error(e);
+        }
+    }
+
+    private static SelectionKey register(
+            final SelectableChannel channel,
+            final Session session,
+            final int ops
+    ) throws IOException {
+        channel.configureBlocking(false);
+        return channel.register(selector, ops, session);
+    }
+
+    @Override
+    public void receiveSession(final Receiver receiver) {
+        final EchoResponse pending = echoResponse.getAndSet(null);
+        if (pending != null) {
+            final ByteBuffer data = receiver.receive(pending.session);
+            assert data != null;
+            data.put(pending.payload);
+            data.flip();
+            return;
+        }
+
+        for (final SelectionKey key : datagramKeys) {
+            if (key.isReadable()) {
+                LOGGER.info("Datagram received");
+                final DatagramChannel channel = (DatagramChannel) key.channel();
+                try {
+                    final DatagramSession session = (DatagramSession) key.attachment();
+                    final ByteBuffer datagram = receiver.receive(session);
+                    assert datagram != null;
+                    final SocketAddress address = channel.receive(datagram);
+                    if (address == null) {
+                        continue;
+                    }
+                    if (Config.useSynchronisedNAT && !address.equals(session.getDestination())) {
+                        continue;
+                    }
+                    datagram.flip();
+                    return;
+                } catch (IOException e) {
+                    LOGGER.error("Trying to read datagram socket", e);
+                }
+            }
+        }
+    }
+
+    @Override
+    public void sendSession(final Session session, @Nullable final ByteBuffer data) {
+        if (session instanceof EchoSession) {
```

Comment:
Can this be solved via polymorphism instead of a type switch? I.e. a `session.send(ByteBuffer, PossiblyNeededContext)` or something?

Reply:
Initially, it was so. However, such an implementation assumes one unnecessary copying of data, and also creates a false impression that the transmitted information is stored in the session object. A session object is an associative object. It is assumed that it is needed in order to distinguish between what part of the information belongs to what. The data itself only exists while the `SessionLayer::sendSession` method is executing. Basically, I keep unpacking the network stack layers even at the session layer. ~~This is more like `ByteBuffer` object contains `Session` object.~~

However, in the case of TCP, the `data` object will indeed be directly stored in the session object. But for the sake of generality and to reduce the number of objects created, I settled on this solution.